File: scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
object CollatzConjecture {
  def steps(n: Int): Option[Int] = n match {
    case x if x <= 0 => None
    case _ => Some(calculate(n))
  }

  private def calculate(n: Int): Int = n match {
    case 1 => 0
    case x if x % 2 == 0 => 1 + calculate(n / 2)
    case x if x % 2 == 1 => 1 + calculate(3 * n + 1)
  }
}

File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let (|Invalid|One|Even|Odd|) n =
    if n <= 0 then Invalid
    elif n = 1 then One
    elif n % 2 = 0 then Even
    else Odd

let inc n = n + 1

let rec steps (number: int): int option =
    match number with
    | Invalid -> None
    | One -> Some 0
    | Even -> steps (number / 2) |> Option.map inc
    | Odd -> steps (3 * number + 1) |> Option.map inc